From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Matti Lehtimaki <matti.lehtimaki@gmail.com>
Date: Mon, 27 Feb 2023 16:49:56 +0200
Subject: [PATCH] (hybris) Revert "Returns a service parse error on overrides
 across the treble boundary."

This reverts commit f597fa5d1dca3b79b30d5b66e15e23b0bd91b560.
---
 init/host_init_verifier.cpp | 85 +++++++------------------------------
 init/service.cpp            |  1 -
 init/service.h              |  2 -
 init/service_parser.cpp     |  8 ----
 init/subcontext.cpp         |  4 --
 init/subcontext.h           |  7 +--
 6 files changed, 17 insertions(+), 90 deletions(-)

diff --git a/init/host_init_verifier.cpp b/init/host_init_verifier.cpp
index 7b8ca91efb6f1ec2d6c3857dec957a5380690d95..3823892e94907f81feb8c100144c8cecc6c86452 100644
--- a/init/host_init_verifier.cpp
+++ b/init/host_init_verifier.cpp
@@ -23,8 +23,6 @@
 #include <fstream>
 #include <iostream>
 #include <iterator>
-#include <map>
-#include <set>
 #include <string>
 #include <vector>
 
@@ -52,7 +50,6 @@
 
 using namespace std::literals;
 
-using android::base::EndsWith;
 using android::base::ParseInt;
 using android::base::ReadFileToString;
 using android::base::Split;
@@ -61,10 +58,6 @@ using android::properties::PropertyInfoEntry;
 
 [[clang::no_destroy]] static std::vector<std::string> passwd_files;
 
-// NOTE: Keep this in sync with the order used by init.cpp LoadBootScripts()
-[[clang::no_destroy]] static const std::vector<std::string> partition_search_order =
-        std::vector<std::string>({"system", "system_ext", "odm", "vendor", "product"});
-
 static std::vector<std::pair<std::string, int>> GetVendorPasswd(const std::string& passwd_file) {
     std::string passwd;
     if (!ReadFileToString(passwd_file, &passwd)) {
@@ -146,24 +139,13 @@ namespace android {
 namespace init {
 
 void PrintUsage() {
-    fprintf(stdout, R"(usage: host_init_verifier [options]
-
-Tests init script(s) for correctness.
-
-Generic options:
-  -p FILE                     Search this passwd file for users and groups.
-  --property_contexts=FILE    Use this file for property_contexts.
-
-Single script mode options:
-  [init rc file]              Positional argument; test this init script.
-
-Multiple script mode options:
-  --out_system=DIR            Path to the output product directory for the system partition.
-  --out_system_ext=DIR        Path to the output product directory for the system_ext partition.
-  --out_odm=DIR               Path to the output product directory for the odm partition.
-  --out_vendor=DIR            Path to the output product directory for the vendor partition.
-  --out_product=DIR           Path to the output product directory for the product partition.
-)");
+    std::cout << "usage: host_init_verifier [options] <init rc file>\n"
+                 "\n"
+                 "Tests an init script for correctness\n"
+                 "\n"
+                 "-p FILE\tSearch this passwd file for users and groups\n"
+                 "--property_contexts=FILE\t Use this file for property_contexts\n"
+              << std::endl;
 }
 
 Result<InterfaceInheritanceHierarchyMap> ReadInterfaceInheritanceHierarchy() {
@@ -210,18 +192,12 @@ int main(int argc, char** argv) {
     android::base::SetMinimumLogSeverity(android::base::ERROR);
 
     auto property_infos = std::vector<PropertyInfoEntry>();
-    std::map<std::string, std::string> partition_map;
 
     while (true) {
         static const char kPropertyContexts[] = "property-contexts=";
         static const struct option long_options[] = {
                 {"help", no_argument, nullptr, 'h'},
                 {kPropertyContexts, required_argument, nullptr, 0},
-                {"out_system", required_argument, nullptr, 0},
-                {"out_system_ext", required_argument, nullptr, 0},
-                {"out_odm", required_argument, nullptr, 0},
-                {"out_vendor", required_argument, nullptr, 0},
-                {"out_product", required_argument, nullptr, 0},
                 {nullptr, 0, nullptr, 0},
         };
 
@@ -237,16 +213,6 @@ int main(int argc, char** argv) {
                 if (long_options[option_index].name == kPropertyContexts) {
                     HandlePropertyContexts(optarg, &property_infos);
                 }
-                for (const auto& p : partition_search_order) {
-                    if (long_options[option_index].name == "out_" + p) {
-                        if (partition_map.find(p) != partition_map.end()) {
-                            PrintUsage();
-                            return EXIT_FAILURE;
-                        }
-                        partition_map[p] =
-                                EndsWith(optarg, "/") ? optarg : std::string(optarg) + "/";
-                    }
-                }
                 break;
             case 'h':
                 PrintUsage();
@@ -263,9 +229,7 @@ int main(int argc, char** argv) {
     argc -= optind;
     argv += optind;
 
-    // If provided, use the partition map to check multiple init rc files.
-    // Otherwise, check a single init rc file.
-    if ((!partition_map.empty() && argc != 0) || (partition_map.empty() && argc != 1)) {
+    if (argc != 1) {
         PrintUsage();
         return EXIT_FAILURE;
     }
@@ -282,43 +246,24 @@ int main(int argc, char** argv) {
         return EXIT_FAILURE;
     }
 
-    if (!partition_map.empty()) {
-        std::vector<std::string> vendor_prefixes;
-        for (const auto& partition : {"vendor", "odm"}) {
-            if (partition_map.find(partition) != partition_map.end()) {
-                vendor_prefixes.push_back(partition_map.at(partition));
-            }
-        }
-        InitializeHostSubcontext(vendor_prefixes);
-    }
-
     const BuiltinFunctionMap& function_map = GetBuiltinFunctionMap();
     Action::set_function_map(&function_map);
     ActionManager& am = ActionManager::GetInstance();
     ServiceList& sl = ServiceList::GetInstance();
     Parser parser;
-    parser.AddSectionParser("service", std::make_unique<ServiceParser>(&sl, GetSubcontext()));
-    parser.AddSectionParser("on", std::make_unique<ActionParser>(&am, GetSubcontext()));
+    parser.AddSectionParser("service", std::make_unique<ServiceParser>(&sl, nullptr));
+    parser.AddSectionParser("on", std::make_unique<ActionParser>(&am, nullptr));
     parser.AddSectionParser("import", std::make_unique<HostImportParser>());
 
-    if (!partition_map.empty()) {
-        for (const auto& p : partition_search_order) {
-            if (partition_map.find(p) != partition_map.end()) {
-                parser.ParseConfig(partition_map.at(p) + "etc/init");
-            }
-        }
-    } else {
-        if (!parser.ParseConfigFileInsecure(*argv, true /* follow_symlinks */)) {
-          // Follow symlinks as inputs during build execution in Bazel's
-          // execution root are symlinks, unlike Soong or Make.
-            LOG(ERROR) << "Failed to open init rc script '" << *argv << "'";
-            return EXIT_FAILURE;
-        }
+    if (!parser.ParseConfigFileInsecure(*argv, true /* follow_symlinks */)) {
+        LOG(ERROR) << "Failed to open init rc script '" << *argv << "'";
+        return EXIT_FAILURE;
     }
 
     size_t failures = parser.parse_error_count() + am.CheckAllCommands() + sl.CheckAllCommands();
     if (failures > 0) {
-        LOG(ERROR) << "Failed to parse init scripts with " << failures << " error(s).";
+        LOG(ERROR) << "Failed to parse init script '" << *argv << "' with " << failures
+                   << " error(s).";
         return EXIT_FAILURE;
     }
 
diff --git a/init/service.cpp b/init/service.cpp
index 3c7b0f6405749c6379e4908e5a95939924e4dd14..0295a8c06568662cfabbd4e20203f565165d9d94 100644
--- a/init/service.cpp
+++ b/init/service.cpp
@@ -168,7 +168,6 @@ Service::Service(const std::string& name, unsigned flags, std::optional<uid_t> u
                  .priority = 0},
       namespaces_{.flags = namespace_flags},
       seclabel_(seclabel),
-      subcontext_(subcontext_for_restart_commands),
       onrestart_(false, subcontext_for_restart_commands, "<Service '" + name + "' onrestart>", 0,
                  "onrestart", {}),
       oom_score_adjust_(DEFAULT_OOM_SCORE_ADJUST),
diff --git a/init/service.h b/init/service.h
index 0095c827062a32260cf7c285973594c9a9b6f33f..1612412734d8920f0d6f784f1559a9ca41e30e66 100644
--- a/init/service.h
+++ b/init/service.h
@@ -151,7 +151,6 @@ class Service {
             flags_ &= ~SVC_ONESHOT;
         }
     }
-    const Subcontext* subcontext() const { return subcontext_; }
     const std::string& filename() const { return filename_; }
     void set_filename(const std::string& name) { filename_ = name; }
     static int GetSigchldFd() {
@@ -204,7 +203,6 @@ class Service {
     // Environment variables that only get applied to the next run.
     std::vector<std::pair<std::string, std::string>> once_environment_vars_;
 
-    const Subcontext* const subcontext_;
     Action onrestart_;  // Commands to execute on restart.
 
     std::vector<std::string> writepid_files_;
diff --git a/init/service_parser.cpp b/init/service_parser.cpp
index 57abbae24bbb4728b8218c7c04fea61e1c3466f6..2971aee1fdfc6cc12d2b4c4baa3af601b278d47d 100644
--- a/init/service_parser.cpp
+++ b/init/service_parser.cpp
@@ -712,14 +712,6 @@ Result<void> ServiceParser::EndSection() {
                            << "' with a config in APEX";
         }
 
-        std::string context = service_->subcontext() ? service_->subcontext()->context() : "";
-        std::string old_context =
-                old_service->subcontext() ? old_service->subcontext()->context() : "";
-        if (context != old_context) {
-            return Error() << "service '" << service_->name() << "' overrides another service "
-                           << "across the treble boundary.";
-        }
-
         service_list_->RemoveService(*old_service);
         old_service = nullptr;
     }
diff --git a/init/subcontext.cpp b/init/subcontext.cpp
index 1b33d3468970b664e5231fe702a3bb202b46f362..ce52502f1572b34b6140f20660bc13d4e027d186 100644
--- a/init/subcontext.cpp
+++ b/init/subcontext.cpp
@@ -363,10 +363,6 @@ void InitializeSubcontext() {
     }
 }
 
-void InitializeHostSubcontext(std::vector<std::string> vendor_prefixes) {
-    subcontext.reset(new Subcontext(vendor_prefixes, {}, kVendorContext, /*host=*/true));
-}
-
 Subcontext* GetSubcontext() {
     return subcontext.get();
 }
diff --git a/init/subcontext.h b/init/subcontext.h
index 23c4a241c98f5623ce68700399ceb8815d6acc99..2fe7770e0b11129064df83258f037efb861cf9e6 100644
--- a/init/subcontext.h
+++ b/init/subcontext.h
@@ -37,14 +37,12 @@ static constexpr const char kTestContext[] = "test-test-test";
 class Subcontext {
   public:
     Subcontext(std::vector<std::string> path_prefixes, std::vector<std::string> partitions,
-               std::string_view context, bool host = false)
+               std::string_view context)
         : path_prefixes_(std::move(path_prefixes)),
           partitions_(std::move(partitions)),
           context_(context.begin(), context.end()),
           pid_(0) {
-        if (!host) {
-            Fork();
-        }
+        Fork();
     }
 
     Result<void> Execute(const std::vector<std::string>& args);
@@ -71,7 +69,6 @@ class Subcontext {
 
 int SubcontextMain(int argc, char** argv, const BuiltinFunctionMap* function_map);
 void InitializeSubcontext();
-void InitializeHostSubcontext(std::vector<std::string> vendor_prefixes);
 Subcontext* GetSubcontext();
 bool SubcontextChildReap(pid_t pid);
 void SubcontextTerminate();
